subsys/bluetooth/host/gatt.c

```diff
@@ -2195,8 +2195,15 @@ static void gatt_ccc_changed(const struct bt_gatt_attr *attr,
 	uint16_t value = 0x0000;
 
 	for (i = 0; i < ARRAY_SIZE(ccc->cfg); i++) {
-		if (ccc->cfg[i].value > value) {
-			value = ccc->cfg[i].value;
+		/* `ccc->value` shall be a summary of connected peers' CCC values, but
+		 * `ccc->cfg` can contain entries for bonded but not connected peers.
+		 */
+		struct bt_conn *conn = bt_conn_lookup_addr_le(ccc->cfg[i].id, &ccc->cfg[i].peer);
+
+		if (conn) {
+			value |= ccc->cfg[i].value;
```

Comment:
I think this has been attempted before. 

Isn't it dangerous to use values that are not defined as a bitfield in the BT core spec as a bitfield?

For example if the BT core spec ever introduces a new value 0x0003 then `value` here would be the same as `BT_GATT_CCC_NOTIFY | BT_GATT_CCC_INDICATE`. 

This change feels less future-proof than the existing, and does not fix any issues. The value here would also be 0x0003 which is not defined in Zephyr, and if applications are testing for `== BT_GATT_CCC_NOTIFY` or `== BT_GATT_CCC_INDICATE` (which we have several cases of in Zephyr), then those checks will break. 

The real issue is that the `cfg_changed` callback does not consider that individual clients have different values. 

```c
	/** @brief CCC attribute changed callback
	 *
	 *  @param attr   The attribute that's changed value
	 *  @param value  New value
	 */
	void (*cfg_changed)(const struct bt_gatt_attr *attr, uint16_t value);
```

should really be
```c
	/** @brief CCC attribute changed callback
	 *
	 *  @param conn   The connection that updated it's CCC value
	 *  @param attr   The attribute that's changed value
	 *  @param value  New value
	 */
	void (*cfg_changed)(struct bt_conn *conn, const struct bt_gatt_attr *attr, uint16_t value);
```

I would suggest to leave the existing callback as is, and introduce the new callback (and possibly deprecate the old one). 

It feels like this is trying to fix a fundamental issue by applying a band-aid, while also introducing a breaking change as the callback value of 0x0003 (BT_GATT_CCC_NOTIFY | BT_GATT_CCC_INDICATE) is undefined.

Reply:
[It is a bit field.](https://www.bluetooth.com/wp-content/uploads/Files/Specification/HTML/Core-60/out/en/host/generic-attribute-profile--gatt-.html#:~:text=Table%203.11:%20Client%20Characteristic%20Configuration-,bit%20field,-definition)

Reply:
> I think this has been attempted before.
> 
> Isn't it dangerous to use values that are not defined as a bitfield in the BT core spec as a bitfield?
> 
> For example if the BT core spec ever introduces a new value 0x0003 then `value` here would be the same as `BT_GATT_CCC_NOTIFY | BT_GATT_CCC_INDICATE`.

But it _is_ defined as a bitfield (Core 6.0):
![image](https://github.com/user-attachments/assets/35f1376d-932d-41e6-90b8-ae493dc4fc4e)

Reply:
I stand corrected :) 

Thanks. 
Should we consider redefining the values in Zephyr from `0x0001` and 0x0002` to `BIT(0)` and `BIT(2)` then?

In Zephyr we have quite a few cases where we treat it not as a bitfield: 
![image](https://github.com/user-attachments/assets/dade7e19-7131-4465-bd4e-e53179839169)
![image](https://github.com/user-attachments/assets/670f6f6a-92f4-4de7-8478-f117a1c2aea2)

They should be updated to use `&` instead of `==` then, as otherwise if the GATT clients subscribe to both indication and notification, then those checks will now fail (granted that all the `== BT_GATT_CCC_NOTIFY` would fail anyways due to the current version using the max of the 2. 

I would also prefer if we _somewhere_ (either in the values or in the documentation for `cfg_changed` or somewhere else) make it clear that these are now bitfields in Zephyr, which they have not been so far. 

Not sure if this change counts as a breaking stable API change, as they have been treated as non-bitfield for 9+ years in Zephyr. I'll leave that up to the maintainers of the host to decide.

Reply:
> Thanks. Should we consider redefining the values in Zephyr from `0x0001` and 0x0002`to`BIT(0)`and`BIT(2)` then?

I agree! And I'll look into switching the checks to bitfields. 
As for the breaking/nonbreaking, we should at least have a release note/changelog. Thoughts @alwa-nordic?

Reply:
> This change feels less future-proof than the existing, and does not fix any issues. The value here would also be 0x0003 which is not defined in Zephyr, and if applications are testing for `== BT_GATT_CCC_NOTIFY` or `== BT_GATT_CCC_INDICATE` (which we have several cases of in Zephyr), then those checks will break.

Yes, this breaks API. But it breaks only in the case when one client enabled indications, and another has enabled notifications. (And the strange case when the same client has enabled both.) The Characteristic Declaration will specify which one of those is supported. Do we have examples of services that support both? Of course there is the issue of a DOS by a rouge client, but the code that does `== BT_GATT_CCC_NOTIFY` will already be vulnerable to this, since we don't check the characteristic properties before we accept the write to the CCC.

> The real issue is that the `cfg_changed` callback does not consider that individual clients have different values. \[We should add `attr` to `cfg_changed`].

Yes, I agree there a need for an event for when `_bt_gatt_ccc.cfg` changes. But this is an orthogonal issue, right? `_bt_gatt_ccc.cfg_changed` is the event for when `_bt_gatt_ccc.value` changes.

This PR instead includes a change to the definition of `_bt_gatt_ccc.value` so it makes sense when a characteristic support both indication and notification. Does this have any real-world uses? I don't know. But the spec allows it, AFAIK.

Reply:
Good examples. In that case, this really is a breaking change. To be on the safe side, we should go with a deprecation, would you agree?

Reply:
We should remove the problematic change from this PR and evaluate how to move forward. I think we should either leave it as it is, or deprecate it without replacement. And we should, as @Thalley points out, address the lack of an event for changes in subscription. The only way to do that without breaking API is to add a new callback.

Reply:
Adding issues:

 - https://github.com/zephyrproject-rtos/zephyr/issues/81004
 - https://github.com/zephyrproject-rtos/zephyr/issues/81005

Reply:
> We should remove the problematic change from this PR and evaluate how to move forward. 

Meaning keeping only the commit that checks whether the peer is connected or not?

Reply:
I added another issue:
 - https://github.com/zephyrproject-rtos/zephyr/issues/81056